azurerm/internal/services/containers/resource_arm_kubernetes_cluster.go

```diff
@@ -424,6 +424,18 @@ func resourceArmKubernetesCluster() *schema.Resource {
 							Computed: true,
 							Elem: &schema.Resource{
 								Schema: map[string]*schema.Schema{
+									"allocated_outbound_port": {
+										Type:         schema.TypeInt,
+										Optional:     true,
+										Default:      -1,
```

Comment:
What is the reason for -1 here? can't we just do d.GetOk() and only pass along the value if it is specified?

Reply:
Nope. Because d.GetOk() will return "0" for TypeInt when you didn't specify "allocated_outbound_port" in tf config. Then service side would set this property as "0" but actually I don't want to set it as "0". I just want to set it as nil.

Reply:
`GetOk` returns a `int, bool`, if the property isn't set i believe the boolean will be false? @neil-yechenwei

Reply:
Yes. The boolean will be false if the property isn't set in tf config. But actually sometimes we don't want to set it as "false". We just want to set it as nil when the property isn't set in tf config.

Reply:
@neil-yechenwei, does this not work:
```
	if port, ok := config["allocated_outbound_port"].(int); ok {
		profile.AllocatedOutboundPorts = utils.Int32(int32(port))
	}
```

Reply:
yes. it doesn't work. Because the code "config["allocated_outbound_port"].(int)" would return "0" when tfconfig doesn't specify the property "allocated_outbound_port" and there is no "Default: -1". Then the property "profile.AllocatedOutboundPorts" would be set as "0" and pass to service side and “0” is meaningful for this property in service side. But actually I just want to set this property as "nil" when tf config doesn't specify this property. So if I use below code, I cannot distinguish user specified "0" or nil.

Reply:
`0` is the default in the service for this setting, and this value has special significance in that it instructs the service to dynamically allocate outbound ports as needed.  I think in this case a default value of `0`  and adding `Computed: true,` here would be acceptable, if properly documented, and exhibit the correct behaviour in use.

Reply:
@jackofallops , I don't think so. If you add `Default: 0` and `Computed: true` for this property, user cannot set it as `nil`. I think we should allow user to not set it. And it would be a breaking change if I update code as you suggested. It would impact other users who don't expect this property. And terraform doesn't allow us to set `Default: 0` and `Computed: true` together on same property. If you do this, it would threw below error message.

property schema:
```
"load_balancer_profile": {
							Type:     schema.TypeList,
							MaxItems: 1,
							ForceNew: true,
							Optional: true,
							Computed: true,
							Elem: &schema.Resource{
								Schema: map[string]*schema.Schema{
									"outbound_ports_allocated": {
										Type:         schema.TypeInt,
										Optional:     true,
										Computed:     true,
										Default:      0,
										ValidateFunc: validation.IntBetween(0, 64000),
									},
```

Error message:
```
            testing.go:640: Step 0 error: config is invalid: Internal validation of the provider failed! This is always a bug
                with the provider itself, and not a user issue. Please report
                this bug:
                
                1 error occurred:
                	* resource azurerm_kubernetes_cluster: outbound_ports_allocated: Default must be nil if computed
                
                
FAIL

Debugger finished with exit code 0

```

Reply:
Sorry - poor wording there. I believe that a read-back value from the API for `AllocatedOutboundPorts` will contain the default value `0`, not `nil`, so we should be fine to set `Computed` and let the zero value pass through?  The user can not set nil explicitly, and omitting the setting should use the default since it must have a value between 0 and 64,000?

Reply:
@jackofallops , sorry, I don't think so. Service API would return `nil` when the property isn't set in tfconfig. So I think user is able to set it as `nil` without `breaking change`. If we set this property must be between 0 and 64000, then breaking change would happen. So I have a concern. When breaking change is allowed in terraform? The breaking change in here is allowed?

tfconfig without `outbound_ports_allocated`:
```
provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "test" {
  name     = "acctestRG-aks-%d"
  location = "westus2"
}

resource "azurerm_kubernetes_cluster" "test" {
  name                = "acctestaks%d"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  dns_prefix          = "acctestaks%d"
  kubernetes_version  = "%s"

  linux_profile {
    admin_username = "acctestuser%d"

    ssh_key {
      key_data = "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCqaZoyiz1qbdOQ8xEf6uEu1cCwYowo5FHtsBhqLoDnnp7KUTEBN+L2NxRIfQ781rxV6Iq5jSav6b2Q8z5KiseOlvKA/RF2wqU0UPYqQviQhLmW6THTpmrv/YkUCuzxDpsH7DUDhZcwySLKVVe0Qm3+5N2Ta6UYH3lsDf9R9wTP2K/+vAnflKebuypNlmocIvakFWoZda18FOmsOoIVXQ8HWFNCuw9ZCunMSN62QGamCe3dL5cXlkgHYv7ekJE15IA9aOJcM7e90oeTqo+7HTcWfdu0qQqPWY5ujyMw/llas8tsXY85LFqRnr3gJ02bAscjc477+X+j/gkpFoN1QEmt terraform@demo.tld"
    }
  }

  default_node_pool {
    name       = "default"
    node_count = 1
    vm_size    = "Standard_DS2_v2"
  }

  service_principal {
    client_id     = "%s"
    client_secret = "%s"
  }

  network_profile {
    network_plugin    = "kubenet"
    load_balancer_sku = "Standard"
    load_balancer_profile {
      managed_outbound_ip_count = 2
    }
  }
}
```

terraform schema:
```
"load_balancer_profile": {
							Type:     schema.TypeList,
							MaxItems: 1,
							ForceNew: true,
							Optional: true,
							Computed: true,
							Elem: &schema.Resource{
								Schema: map[string]*schema.Schema{
									"outbound_ports_allocated": {
										Type:         schema.TypeInt,
										Optional:     true,
										Default:      0,
										ValidateFunc: validation.IntBetween(0, 64000),
									},
```

The result of `outbound_ports_allocated` from service API:
![image](https://user-images.githubusercontent.com/19754191/77128295-881f3880-6a8a-11ea-9842-8938d449ea7d.png)

Reply:
Hi @neil-yechenwei 
I meant the following for the schema:
```
						"load_balancer_profile": {
							Type:     schema.TypeList,
							MaxItems: 1,
							ForceNew: true,
							Optional: true,
							Computed: true,
							Elem: &schema.Resource{
								Schema: map[string]*schema.Schema{
									"outbound_ports_allocated": {
										Type:         schema.TypeInt,
										Optional:     true,
										Computed:     true,
										ValidateFunc: validation.IntBetween(0, 64000),
									},
```
And can I see the read comes back with a `0` from the API:
![image](https://user-images.githubusercontent.com/11830746/77142007-b6931880-6a76-11ea-9240-1ece46dc1fc7.png)

![image](https://user-images.githubusercontent.com/11830746/77142253-6ec0c100-6a77-11ea-87e7-2e9f7c19edfe.png)


with config:
```
 network_profile {
    network_plugin    = "kubenet"
    load_balancer_sku = "Standard"
    load_balancer_profile {
      managed_outbound_ip_count = 2
      idle_timeout_in_minutes   = 10
    }
```
Does that help explain what I mean?

Reply:
@jackofallops , sorry, seems I still cannot get your point. I still think your change would cause `breaking change` since some user maybe doesn't expect this change would impact their existing service. So I am asking does terraform allow `breaking change` in this situation. If yes, I can update this PR to `force user to set this property as 0-64000`. So do you mean ` user must set property "outbound_ports_allocated" as 0~64000`? And I've tried your above changes, you actually already set this property `outbound_ports_allocated` as `0` before calling service API so that you would get `0` from service side. Actually service API would return `nil` as default. Please see my below example. If you think I am not right, could you commit your whole changes to this PR for understanding?

tfconfig without property `outbound_ports_allocated`:
```
provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "test" {
  name     = "acctestRG-aks-%d"
  location = "westus2"
}

resource "azurerm_kubernetes_cluster" "test" {
  name                = "acctestaks%d"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  dns_prefix          = "acctestaks%d"
  kubernetes_version  = "%s"

  linux_profile {
    admin_username = "acctestuser%d"

    ssh_key {
      key_data = "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCqaZoyiz1qbdOQ8xEf6uEu1cCwYowo5FHtsBhqLoDnnp7KUTEBN+L2NxRIfQ781rxV6Iq5jSav6b2Q8z5KiseOlvKA/RF2wqU0UPYqQviQhLmW6THTpmrv/YkUCuzxDpsH7DUDhZcwySLKVVe0Qm3+5N2Ta6UYH3lsDf9R9wTP2K/+vAnflKebuypNlmocIvakFWoZda18FOmsOoIVXQ8HWFNCuw9ZCunMSN62QGamCe3dL5cXlkgHYv7ekJE15IA9aOJcM7e90oeTqo+7HTcWfdu0qQqPWY5ujyMw/llas8tsXY85LFqRnr3gJ02bAscjc477+X+j/gkpFoN1QEmt terraform@demo.tld"
    }
  }

  default_node_pool {
    name       = "default"
    node_count = 1
    vm_size    = "Standard_DS2_v2"
  }

  service_principal {
    client_id     = "%s"
    client_secret = "%s"
  }

  network_profile {
    network_plugin    = "kubenet"
    load_balancer_sku = "Standard"
    load_balancer_profile {
      managed_outbound_ip_count = 2
    }
  }
}
```

terraform schema:
```
"load_balancer_profile": {
							Type:     schema.TypeList,
							MaxItems: 1,
							ForceNew: true,
							Optional: true,
							Computed: true,
							Elem: &schema.Resource{
								Schema: map[string]*schema.Schema{
									"outbound_ports_allocated": {
										Type:         schema.TypeInt,
										Optional:     true,
										Computed:     true,
										ValidateFunc: validation.IntBetween(0, 64000),
									},
```

Do not set property `outbound_ports_allocated` before calling service api:
![image](https://user-images.githubusercontent.com/19754191/77220925-e24df580-6b7f-11ea-932c-b5f4d8ef0e27.png)

Getting value of property `outbound_ports_allocated` from Service API:
![image](https://user-images.githubusercontent.com/19754191/77220935-0b6e8600-6b80-11ea-824b-98a12c139d0b.png)

So service API would return `nil` NOT `0` when you didn't set this property in tfconfig.

Reply:
@jackofallops , I've double confirmed with this service owner. They said property "AllocatedOutboundPorts" would be set to “0” and property “IdleTimeoutInMinutes” would be set to “30” in service side when we don't set these two properties in request body. So I've updated code. Please have a look. Thanks.

The answer from service owner:
Today, if a customer doesn’t set them explicitly then we don’t fill the two properties from the AKS side but they do get filled on the SLB side with the defaults.